batcher/sweep_batch.go

```diff
@@ -0,0 +1,1109 @@
+		blockEpochChan:   make(chan int32),
+		spendChan:        make(chan *chainntnfs.SpendDetail),
+		spendRefreshChan: make(chan int32, 1),
+		confChan:         make(chan *chainntnfs.TxConfirmation),
+		confRefreshChan:  make(chan int32, 1),
+		errChan:          make(chan error),
+		completeChan:     completeChan,
+		BatchTxid:        batchTxid,
+		wallet:           wallet,
+		chainNotifier:    chainNotifier,
+		signerClient:     signerClient,
+		muSig2SignSweep:  musig2SignSweep,
+		store:            store,
+		Cfg:              &cfg,
+	}, nil
+}
+
+// NewBatch creates a new batch that already existed in storage.
+func NewBatchFromDB(cfg BatchConfig, ID lntypes.Hash, batchTxid *chainhash.Hash,
+	batchPkScript []byte, state batchState, primaryID lntypes.Hash,
+	sweeps map[lntypes.Hash]Sweep, rbfCache RBFCache,
+	returnChan chan *Sweep, completeChan chan lntypes.Hash,
+	wallet lndclient.WalletKitClient,
+	chainNotifier lndclient.ChainNotifierClient,
+	signerClient lndclient.SignerClient, musig2SignSweep MuSig2SignSweep,
+	store BatcherStore) (*Batch, error) {
+
+	newID := lntypes.Hash{}
+	_, err := crand.Read(newID[:])
+	if err != nil {
+		return nil, err
+	}
+
+	return &Batch{
+		ID:               ID,
+		State:            state,
+		BatchReady:       false,
+		PrimarySweepID:   primaryID,
+		Sweeps:           sweeps,
+		RequestChan:      make(chan *Sweep),
+		returnChan:       returnChan,
+		blockEpochChan:   make(chan int32),
+		spendChan:        make(chan *chainntnfs.SpendDetail),
+		spendRefreshChan: make(chan int32, 1),
+		confChan:         make(chan *chainntnfs.TxConfirmation),
+		confRefreshChan:  make(chan int32, 1),
+		errChan:          make(chan error),
+		completeChan:     completeChan,
+		BatchTxid:        batchTxid,
+		BatchPkScript:    batchPkScript,
+		RbfCache:         rbfCache,
+		wallet:           wallet,
+		chainNotifier:    chainNotifier,
+		signerClient:     signerClient,
+		muSig2SignSweep:  musig2SignSweep,
+		store:            store,
+		Cfg:              &cfg,
+	}, nil
+}
+
+// addSweep adds a sweep to the batch. If this is the first sweep being added
+// to the batch then it also sets the primary sweep ID.
+func (b *Batch) addSweep(sweep Sweep) {
+
+	if b.PrimarySweepID == lntypes.ZeroHash {
+		b.PrimarySweepID = sweep.SwapHash
+		b.BatchReady = true
+		b.Cfg.BatchConfTarget = sweep.ConfTarget
+	}
+
+	if b.PrimarySweepID == sweep.SwapHash {
+		b.BatchReady = true
+		b.Cfg.BatchConfTarget = sweep.ConfTarget
+	}
+
+	b.Sweeps[sweep.SwapHash] = sweep
+}
+
+// SweepExists returns true if the batch contains the sweep with the given hash.
+func (b *Batch) SweepExists(hash lntypes.Hash) bool {
+	_, ok := b.Sweeps[hash]
+	return ok
+}
+
+// AcceptsSweep returns true if the batch is able to accept the given sweep.
+func (b *Batch) AcceptsSweep(sweep *Sweep) bool {
+	tooFar := false
+	for _, s := range b.Sweeps {
+		if int32(math.Abs(float64(sweep.Timeout-s.Timeout))) >
+			b.Cfg.MaxTimeoutDistance {
+
+			tooFar = true
```

Comment:
nit: can return false here

Reply:
That would be valid. I wonder if keeping it all in a single line like this
```
return (b.State == Open) && !tooFar
```
expresses the conditions in a more visible way?

Reply:
Yes true. What we should still do is check if `sweep` is nil and also simplify the `if` clause. I have:
```
func (b *Batch) AcceptsSweep(sweep *Sweep) bool {
	if sweep == nil {
		return false
	}

	tooFar := false
	for _, s := range b.Sweeps {
		timeoutDistance := int32(math.Abs(float64(sweep.Timeout - s.Timeout)))

		if timeoutDistance > b.Cfg.MaxTimeoutDistance {
			tooFar = true
		}
	}

	return (b.State == Open) && !tooFar
}
```